### ISSUE TITLE
[Bug] Alert with divideSeries error, but idential 2nd alert works.

### ISSUE THREAD

**Issue by wleese** [grafana/grafana#7738]
Please include this information:
- What Grafana version are you using?
4.1.2

Created alert for 2 identical metrics. One has a working alert and the other one fails with a graphite internal error.

```
{
  "__inputs": [
    {
      "name": "DS_GRAPHITE",
      "label": "graphite",
      "description": "",
      "type": "datasource",
      "pluginId": "graphite",
      "pluginName": "Graphite"
    }
  ],
  "__requires": [
    {
      "type": "grafana",
      "id": "grafana",
      "name": "Grafana",
      "version": "4.1.2"
    },
    {
      "type": "panel",
      "id": "graph",
      "name": "Graph",
      "version": ""
    },
    {
      "type": "datasource",
      "id": "graphite",
      "name": "Graphite",
      "version": "1.0.0"
    }
  ],
  "annotations": {
    "list": []
  },
  "editMode": false,
  "editable": true,
  "gnetId": null,
  "graphTooltip": 1,
  "hideControls": false,
  "id": null,
  "links": [],
  "refresh": "1m",
  "rows": [
    {
      "collapse": false,
      "height": 250,
      "panels": [
        {
          "alert": {
            "conditions": [
              {
                "evaluator": {
                  "params": [
                    20
                  ],
                  "type": "gt"
                },
                "operator": {
                  "type": "and"
                },
                "query": {
                  "params": [
                    "E",
                    "5m",
                    "now"
                  ]
                },
                "reducer": {
                  "params": [],
                  "type": "avg"
                },
                "type": "query"
              }
            ],
            "executionErrorState": "alerting",
            "frequency": "60s",
            "handler": 1,
            "name": "Fail built measurement rate alert",
            "noDataState": "no_data",
            "notifications": []
          },
          "aliasColors": {},
          "bars": false,
          "datasource": "${DS_GRAPHITE}",
          "fill": 1,
          "id": 51,
          "legend": {
            "alignAsTable": true,
            "avg": true,
            "current": true,
            "max": true,
            "min": false,
            "show": true,
            "total": false,
            "values": true
          },
          "lines": true,
          "linewidth": 1,
          "links": [],
          "nullPointMode": "null as zero",
          "percentage": false,
          "pointradius": 5,
          "points": false,
          "renderer": "flot",
          "seriesOverrides": [],
          "span": 12,
          "stack": false,
          "steppedLine": false,
          "targets": [
            {
              "hide": true,
              "refId": "A",
              "target": "groupByNode(statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.failBuiltMeasurementAmount.m1_rate, 3, 'sum')",
              "textEditor": true
            },
            {
              "hide": true,
              "refId": "B",
              "target": "groupByNode(statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.collectedMeasurementAmount.m1_rate, 3, 'sum')",
              "textEditor": true
            },
            {
              "hide": true,
              "refId": "E",
              "target": "divideSeries(#A,sumSeries(statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.failBuiltMeasurementAmount.m1_rate,statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.collectedMeasurementAmount.m1_rate))",
              "targetFull": "alias(scale(divideSeries(groupByNode(statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.failBuiltMeasurementAmount.m1_rate, 3, 'sum'), sumSeries(#A, #B)), 100), 'Error rate')",
              "textEditor": true
            },
            {
              "refId": "F",
              "target": "divideSeries(#A,sumSeries(statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.failBuiltMeasurementAmount.m1_rate,statsd.pro.*.wsp.*-wsp-app-*.ps*.m2.core.collector.collectedMeasurementAmount.m1_rate))",
              "textEditor": true
            }
          ],
          "thresholds": [
            {
              "colorMode": "critical",
              "fill": true,
              "line": true,
              "op": "gt",
              "value": 20
            }
          ],
          "timeFrom": null,
          "timeShift": null,
          "title": "Fail built measurement rate",
          "tooltip": {
            "shared": true,
            "sort": 0,
            "value_type": "individual"
          },
          "type": "graph",
          "xaxis": {
            "mode": "time",
            "name": null,
            "show": true,
            "values": []
          },
          "yaxes": [
            {
              "format": "percent",
              "label": null,
              "logBase": 1,
              "max": "100",
              "min": "0",
              "show": true
            },
            {
              "format": "short",
              "label": null,
              "logBase": 1,
              "max": null,
              "min": null,
              "show": false
            }
          ]
        }
      ],
      "repeat": null,
      "repeatIteration": null,
      "repeatRowId": null,
      "showTitle": false,
      "title": "Dashboard Row",
      "titleSize": "h6"
    }
  ],
  "schemaVersion": 14,
  "style": "dark",
  "tags": [
    "teamm2_lg_pro"
  ],
  "templating": {
    "list": [
      {
        "allValue": null,
        "current": {},
        "datasource": "${DS_GRAPHITE}",
        "hide": 0,
        "includeAll": false,
        "label": null,
        "multi": false,
        "name": "env",
        "options": [],
        "query": "statsd.*",
        "refresh": 1,
        "regex": "",
        "sort": 0,
        "tagValuesQuery": null,
        "tags": [],
        "tagsQuery": null,
        "type": "query",
        "useTags": false
      }
    ]
  },
  "time": {
    "from": "now-7d",
    "to": "now"
  },
  "timepicker": {
    "refresh_intervals": [
      "5s",
      "10s",
      "30s",
      "1m",
      "5m",
      "15m",
      "30m",
      "1h",
      "2h",
      "1d"
    ],
    "time_options": [
      "5m",
      "15m",
      "1h",
      "6h",
      "12h",
      "24h",
      "2d",
      "7d",
      "30d"
    ]
  },
  "timezone": "browser",
  "title": "William test",
  "version": 3
}
```

In the above dashboard, an alert on query E works:

```
result

    firing false
    conditionEvals false = false
    timeMs -147.821ms
    logs
        0
            message Condition[0]: Query Result
            data
```

An alert on identical query F fails:

```
 result

    firing true
    conditionEvals = true
    timeMs 70.166ms
    error tsdb.HandleRequest() error Request failed status: 500 INTERNAL SERVER ERROR

t=2017-03-06T10:01:05+0100 lvl=info msg="Request failed" logger=tsdb.graphite status="500 INTERNAL SERVER ERROR" body="<body style=\"background-color: #666666; color: black;\">\n<center>\n<h2 style='font-family: \"Arial\"'>\n<p>Graphite encountered an unexpected error while handling your request.</p>\n<p>Please contact your site administrator if the problem persists.</p>\n</h2>\n<br/>\n<div style=\"width: 50%; text-align: center; font-family: monospace; background-color: black; font-weight: bold; color: #ff4422;\">\n\n</div>\n\n<div style=\"width: 70%; text-align: left; background-color: black; color: #44ff22; border: thin solid gray;\">\n<pre>\nTraceback (most recent call last):\n  File &quot;/usr/lib/python2.6/site-packages/django/core/handlers/base.py&quot;, line 109, in get_response\n    response = callback(request, *callback_args, **callback_kwargs)\n  File &quot;/opt/graphite/webapp/graphite/render/views.py&quot;, line 107, in renderView\n    seriesList = evaluateTarget(requestContext, target)\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 10, in evaluateTarget\n    result = evaluateTokens(requestContext, tokens)\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 21, in evaluateTokens\n    return evaluateTokens(requestContext, tokens.expression)\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 28, in evaluateTokens\n    args = [evaluateTokens(requestContext, arg) for arg in tokens.call.args]\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 21, in evaluateTokens\n    return evaluateTokens(requestContext, tokens.expression)\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 28, in evaluateTokens\n    args = [evaluateTokens(requestContext, arg) for arg in tokens.call.args]\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 21, in evaluateTokens\n    return evaluateTokens(requestContext, tokens.expression)\n  File &quot;/opt/graphite/webapp/graphite/render/evaluator.py&quot;, line 29, in evaluateTokens\n    return func(requestContext, *args)\n  File &quot;/opt/graphite/webapp/graphite/render/functions.py&quot;, line 473, in divideSeries\n    raise ValueError(&quot;divideSeries second argument must reference exactly 1 series&quot;)\nValueError: divideSeries second argument must reference exactly 1 series\n\n</pre>\n</div>\n\n</center>\n"
t=2017-03-06T10:01:05+0100 lvl=info msg="Request Completed" logger=context userId=14 orgId=1 uname="William Leese" method=POST path=/api/alerts/test status=200 remote_addr=10xxxxxxx time_ms=345ns size=150
```

**Comment by torkelo** [comment]
query F is missing the targetFull property (in the json domain model). This is something the frontend generates when composing nested queries using `#<refId>` syntax. 

try updating this query (simple change divideSeries(#A) parameter to #B, and then back to #C) this will cause grafana to update/add the targetFull property (new and needed for alerting)

**Comment by wleese** [comment]
> query F is missing the targetFull property (in the json domain model). This is something the frontend generates when composing nested queries using #<refId> syntax.

So the frontend has a bug? Trying to understand why this ticket is closed :)

**Comment by torkelo** [comment]
It is not a bug. But a migration problem. Queries that use other queries (in a nested way using syntax `#<refid`) are a problem in alerting as altering only gets access to the query model for the specified letter. 

So to make existing queries (created in dashboards before v4.) usable in alerting you just need to change something in that query row (then back). Then save the dashboard.

**Comment by wleese** [comment]
I performed the migration from 3.x to 4.1.

Did I miss a step, got the upgrade ordering wrong or was there simply no migration done for this case?

**Comment by torkelo** [comment]
there is no migration for this step. The migration is too complicated and costly. 

Not many has run into it as most create new panels specifically for alerts, only occurs on old panels with nested queries.

**Comment by wleese** [comment]
Assumptions on how people use grafana ... these are the kind of situations that really hurt big deployments. 

As a sysadmin serving 300+ users with 1000+ dashboards, knowing that things are dealt with this way makes me allot more nervous for the next big grafana release.

**Comment by torkelo** [comment]
I am sorry you feel that way. We tried to look into doing a migration for this but proved impossible. 

But we should probably have a specific error message message for this. It was a tricky case, only graphite queries can refer to other queries, and the alert condition only gets to know the single query model specifed into the query condition. So we had to update how nested queries work so they update the query model with a full query. 

So we should handled this better. But it's not like a hidden bug that will bite users in unexpected ways as the alert rule will always fail when you test it.
